### PR TITLE
Add resource package prefixes

### PR DESCRIPTION
**Pull request by z-adams**
Adds the resource package prefixes Cap planned originally.

I recommend looking at Package.h/Package.cpp first, then OSPApplication.h/.cpp. The rest of the changes are simply updating other code to use the new system. I wish there were a way to order the files in the "files changed" tab so that the logical cascade of changes was easier to follow.

### CODE REVIEW

File: src/osp/OSPApplication.cpp
Comment (jonesmz) [inline src/osp/OSPApplication.cpp:34]: Careful with assuming that .emplace() will succeed.

1) Functions that just call a function on a member variable should either A) Return what the called function returns (Or a part of it, like the success / fail) or B) should assert at runtime that the inner call succeeded.
2) Do you really want to accept by rvalue-reference only? If so, that's fine, but idomatically most people assume you'll accept a copy as well as a move. If you need it to be rvalue-reference only, consider explicitly deleting the copy constructor and assignment operator off of the Package class.
3) There's a good chance that you want try_emplace over emplace, because try_emplace only actually move constructs the provided value if-and-only-if the insert operation will succeed.
4) If you don't want try_emplace, and you don't want to assert / return the bool, then maybe you want insert_or_assign to overwrite the existing key (if exists)

File: src/osp/OSPApplication.cpp
Comment (jonesmz) [inline src/osp/OSPApplication.cpp:39]: resprefix_t is typedefed as std::string

but here you're taking it by value for use with a function that doesn't require ownership semantics.

It's far far more likely that what you want to do here is take an std::string_view (Give it an appropriate typedef if you want), and then use std::map::find(), and if the result of find() == std::map::end(), then throw the exception that std::map::at() would throw.

If you don't want to use string_view (I strongly recommend it for this situation), then at the least, take the argument as const&

File: src/osp/OSPApplication.h
Comment (jonesmz) [inline src/osp/OSPApplication.h:43]: Remove commented out code.

File: src/osp/OSPApplication.h
Comment (jonesmz) [inline src/osp/OSPApplication.h:68]: std::map<resprefix_t, Package, std::less<>>

File: src/osp/Resource/Package.cpp
Comment (jonesmz) [inline src/osp/Resource/Package.cpp:32]: Most likely you want these constructor parameters to be passed in by value and then moved, not by const&.

File: src/osp/Resource/Package.cpp
Comment (jonesmz) [inline src/osp/Resource/Package.cpp:40]: std::string_view. The caller of the function may not inherently have an std::string that they can pass in. They might have some other string type.

File: src/osp/Resource/Package.h
Comment (jonesmz) [inline src/osp/Resource/Package.h:65]: Use the using keyword over typedef. Typedef sucks.

File: src/osp/Resource/Package.h
Comment (jonesmz) [inline src/osp/Resource/Package.h:68]: There's no reason to make this a typedef at all. Referencing it by the name of the struct type is perfectly fine.

Comment (jonesmz) [reply src/osp/Resource/Package.h:68]: Also, man, that's a lot of string copying. You sure? I feel like what you want here is actually two std::string_views and then a single string object that owns the storage (Careful with copy / move constructors though).

Or maybe even more accurately, you might want something like an std::string_range (doesn't exist, far as I know) that's basically just two size_t's that indicate the range of indicies in some std::string, applied as needed to get back an std::string_view.

File: src/osp/Resource/Package.h
Comment (jonesmz) [inline src/osp/Resource/Package.h:65]: I may be misremembering, but I thought the project's coding guidelines document encouraged capitalizing the first letter in the typedef name?

**Review by jonesmz — APPROVED** [review]

File: src/osp/Resource/Package.h
Comment (Capital-Asterisk) [reply src/osp/Resource/Package.h:68]: Very C-style indeed. There's many ways to go upon doing this, depending on how this struct will be used. Will this struct own a string or is it just a return type for decompose_path?

File: src/osp/Resource/Package.cpp
Comment (Capital-Asterisk) [reply src/osp/Resource/Package.cpp:40]: This can be turned into a more generic string_view parsing function: inputting a string view and returning a pair of string views. If Path is intended to own a string value, then the pair can be used to initialize a Path.

**Review by Capital-Asterisk — APPROVED** [review]

File: src/osp/Active/SysVehicle.cpp
Comment (z-adams) [reply src/osp/Active/SysVehicle.cpp:285]: I'll change it to "find," it conveys the behavior more clearly

File: src/osp/OSPApplication.h
Comment (z-adams) [reply src/osp/OSPApplication.h:68]: Is this not the default behavior?

File: src/osp/Resource/Package.h
Comment (z-adams) [reply src/osp/Resource/Package.h:65]: The primary reason why I did that was so that `OSPApplication::m_packages` was a `std::map<resprefix_t, Package>` which makes it more obvious that the key is not just any old string but rather a specific engine construct. Sort of like how `size_t` (which is just `uint64_t`) conveys a specific meaning. I'll leave it up to you whether or not I should leave it this way.

File: src/osp/Resource/Package.cpp
Comment (z-adams) [reply src/osp/Resource/Package.cpp:40]: I can do that, although I'll probably end up making another `decompose_path` function that just wraps that one. The purpose of this one is to make it easy to parse resource names. I just realized that this is not used anywhere right now, but it will be once the rest of my branch is merged.

File: src/osp/OSPApplication.h
Comment (Capital-Asterisk) [reply src/osp/OSPApplication.h:68]: This "transparent comparator" allows string views to access the map. If I understand correctly, it will allow any type comparable to an std::string.

File: src/osp/Resource/Package.h
Comment (z-adams) [reply src/osp/Resource/Package.h:68]: Can't remember how it ended up this way, I suspect I originally typedef'd a `std::pair` or something and then forgot to get rid of it when I changed to a struct.

File: src/osp/OSPApplication.cpp
Comment (z-adams) [reply src/osp/OSPApplication.cpp:34]: We do indeed want move-only behavior, the Package class's copy constructor is already deleted (although the copy assign is missing, I'll delete that one too). I don't think it makes sense to allow the same key to be added multiple times so it should probably just fail. I'll assert for now; if @Capital-Asterisk thinks overwriting is better I can change it.

File: src/osp/Resource/Package.h
Comment (z-adams) [reply src/osp/Resource/Package.h:68]: Looks like string_view will work pretty much everywhere, I will either just change everything to string_view or make an equivalent typedef for the resprefix, depending on whether cap thinks it's reasonable to use the typedef in the first place.

**Comment by jonesmz** [comment]
I'm fine with merging this, so whenever you're happy @Capital-Asterisk

**Review by Capital-Asterisk — APPROVED** [review]

**Pull request merged**
